java/ql/lib/semmle/code/java/frameworks/ratpack/RatpackExec.qll

```diff
@@ -46,12 +49,57 @@ private class RatpackExecModel extends SummaryModelCsv {
           "flatMap;;;Element of Argument[-1];Parameter[0] of Argument[0];value",
           "flatMap;;;Element of ReturnValue of Argument[0];Element of ReturnValue;value",
           "flatMapError;;;Element of ReturnValue of Argument[1];Element of ReturnValue;value",
+          // `blockingOp` passes the value to the argumet
+          "blockingOp;;;Element of Argument[-1];Parameter[0] of Argument[0];value",
+          // `replace` returns the passed `Promise`
+          "replace;;;Argument[0];ReturnValue;value",
```

Comment:
The doc and the source both read like you get a new promise back, that discards the outcome of `this` and replaces it with the outcome of arg0?

Reply:
Not quite, it's not discarding the promise, because everything earlier in the chain still needs to execute.

Here's how the method is implemented on the interface:
```java
  /**
   * Replaces {@code this} promise with the provided promise for downstream subscribers.
   * <p>
   * This is simply a more convenient form of {@link #flatMap(Function)}, where the given promise is returned.
   * This method can be used when a subsequent operation on a promise isn't dependent on the actual promised value.
   * <p>
   * If the upstream promise fails, its error will propagate downstream and the given promise will never be subscribed to.
   *
   * <pre class="java">{@code
   * import ratpack.test.exec.ExecHarness;
   * import ratpack.exec.ExecResult;
   * import ratpack.exec.Promise;
   *
   * import static org.junit.Assert.assertEquals;
   *
   * public class Example {
   *   private static String value;
   *
   *   public static void main(String... args) throws Exception {
   *     ExecResult<String> result = ExecHarness.yieldSingle(c ->
   *         Promise.value("foo")
   *           .next(v -> value = v)
   *           .replace(Promise.value("bar"))
   *     );
   *
   *     assertEquals("bar", result.getValue());
   *     assertEquals("foo", value);
   *   }
   * }
   * }</pre>
   *
   * @param next the promise to replace {@code this} with
   * @param <O> the type of the value of the replacement promise
   * @return a promise
   * @since 1.1
   */
  default <O> Promise<O> replace(Promise<O> next) {
    return flatMap(in -> next);
  }
```

https://github.com/ratpack/ratpack/blob/1a5d97a011930335fdc97b20a638b6afb070a4a4/ratpack-exec/src/main/java/ratpack/exec/Promise.java#L767-L805

Reply:
Right, but that's a fresh object, so we don't have value-flow from `next` to the return value

Reply:
I mean, sure, but the semantics are the same from a CodeQL perspective, are they not?

Reply:
Generally we would call aliases value-flow and anything else taint-flow

Reply:
Okay, I've updated this to be `replace;;;Element of Argument[0];Element of ReturnValue;value`